KAFKA-8806 Reduce calls to validateOffsetsIfNeeded

mumrah opened the pull request:
Remember the `updateVersion` of the last update to Metadata so we can avoid unnecessarily checking each partition for leader epoch changes on every call to `KafkaConsumer#poll`

mumrah committed:
Only check for leader changes when there is new metadata

mumrah committed:
Add missing import

mumrah committed:
checkstyle

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -479,10 +481,14 @@ public void validateOffsetsIfNeeded() {
-        });
+        // If we see a new metadata version, check all partitions
+        int newMetadataUpdateVersion = metadata.updateVersion();
+        if (metadataUpdateVersion.getAndSet(newMetadataUpdateVersion) != newMetadataUpdateVersion) {
```
Since we have the same logic in two places, maybe it makes sense to have a method?

mumrah replied:
done

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -1091,9 +1098,12 @@ Node selectReadReplica(TopicPartition partition, Node leaderReplica, long curren
     private Map<Node, FetchSessionHandler.FetchRequestData> prepareFetchRequests() {
```
The `fetchablePartitions` method used below is probably another nice opportunity to use something like `forEachAssignedPartition`.

mumrah replied:
I'm not sure it saves anything in this case. The main benefit of `forEachAssignedPartition` is avoiding making a copy of the assignment set. Since `fetchablePartitions` iterates across the internal set directly I don't think it would help

ijuma replied:
We can avoid a copy in the case @hachikuji mentions as well, right? See below:

```java
synchronized List<TopicPartition> fetchablePartitions(Predicate<TopicPartition> isAvailable) {
        return assignment.stream()
                .filter(tpState -> isAvailable.test(tpState.topicPartition()) && tpState.value().isFetchable())
                .map(PartitionStates.PartitionState::topicPartition)
                .collect(Collectors.toList());
    }
```

mumrah replied:
Ah ok, I see what he means. I'll look into this

mumrah replied:
I made a pass at this and it wasn't so simple. Deferring for now

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -388,6 +390,10 @@ synchronized void maybeSeekUnvalidated(TopicPartition tp, long offset, OffsetRes
         return new HashSet<>(this.assignment.partitionSet());
     }
 
+    public void forEachAssignedPartition(Consumer<TopicPartition> consumer) {
```
Should this be synchronized?

mumrah replied:
yes, thanks

hachikuji reviewed (commented):
Thanks, left a couple comments.

mumrah committed:
Feedback from PR

mumrah committed:
checkstyle

ijuma commented:
There are a number of test failures that could be related.

mumrah commented:
@ijuma, they are related, looking at them now.

mumrah committed:
Fix FetcherTest tests

mumrah committed:
Use full FetchPosition when resetting position

Previouslly, this would only update the offset and rely on future calls to
Fetcher#maybeValidatePositionForCurrentLeader to get the leader information.
Now that we are only calling maybeValidatePositionForCurrentLeader when the
metadata has updated, we would get stuck after a reset.

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8806-reduce-metadata-update-calls

mumrah committed:
Fix mocking in unit test

hachikuji reviewed (approved):
Thanks, LGTM. I'll merge after the build completes.

mumrah commented:
retest this please

ijuma commented:
Two jobs timed out, one had flaky test failures:

> kafka.api.GroupEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl
> kafka.api.SaslSslAdminClientIntegrationTest.testDescribeLogDirs
> kafka.api.SaslSslAdminClientIntegrationTest.testAlterReplicaLogDirs
> org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testReconfigConnector
> kafka.api.DelegationTokenEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl
>

ijuma commented:
retest this please

hachikuji commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -711,7 +711,7 @@ private void resetOffsetIfNeeded(TopicPartition partition, OffsetResetStrategy r
                 offsetData.offset, offsetData.leaderEpoch, metadata.leaderAndEpoch(partition));
         offsetData.leaderEpoch.ifPresent(epoch -> metadata.updateLastSeenEpochIfNewer(partition, epoch));
-        subscriptions.maybeSeekUnvalidated(partition, position.offset, requestedResetStrategy);
+        subscriptions.maybeSeekUnvalidated(partition, position, requestedResetStrategy);
```
Mentioned offline, but there is actually a behavioral change here. Previously we didn't validate after an offset reset in any case, but now I think we might if the offset returned from the ListOffset request has an epoch. I think the simple thing to do is to use `Optional.empty()` above when defining `position`.

mumrah replied:
+1 for not introducing a behavior change here. Fixed. Let's see how the next build goes.

mumrah committed:
Ensure we don't enter validation when reseting offsets

mumrah commented:
retest this please

hachikuji commented:
retest this please

ijuma commented:
retest this please

ijuma commented:
It's a bit concerning that the tests are so flaky in this PR, have we been checking the failures to see if they're related (before Jenkins deletes them)?

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8806-reduce-metadata-update-calls

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8806-reduce-metadata-update-calls

mumrah committed:
Temporarily revert back to always checking the position

mumrah committed:
Adding verbose logging

mumrah committed:
checkstyle

hachikuji commented:
retest this please

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-8806-reduce-metadata-update-calls

mumrah committed:
Fixup after merge

mumrah commented:
Reviving this PR cc @hachikuji @ijuma @andrewchoi5

mumrah commented:
![image](https://user-images.githubusercontent.com/55116/89447599-3e2aac80-d724-11ea-8d0d-a4a4857f230c.png)

Failed run https://builds.apache.org/job/kafka-pr-jdk14-scala2.13/1860/

mumrah commented:
retest this please

mumrah commented:
![image](https://user-images.githubusercontent.com/55116/89474075-8f519500-d752-11ea-9853-1bd202a5c208.png)

mumrah commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -1142,6 +1156,7 @@ Node selectReadReplica(TopicPartition partition, Node leaderReplica, long curren
 
             Optional<Node> leaderOpt = position.currentLeader.leader;
             if (!leaderOpt.isPresent()) {
+                log.info("Requesting metadata update for partition {} since the position {} is missing the current leader node", partition, position);
```
This feels more like a debug level log to me. Was there a specific reason we needed it?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -406,6 +407,10 @@ synchronized void maybeSeekUnvalidated(TopicPartition tp, long offset, OffsetRes
         return new HashSet<>(this.assignment.partitionSet());
     }
 
+    public synchronized void forEachAssignedPartition(Consumer<TopicPartition> consumer) {
```
This is unused

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -1120,17 +1123,28 @@ Node selectReadReplica(TopicPartition partition, Node leaderReplica, long curren
+     * If we have seen new metadata (as tracked by {@link org.apache.kafka.clients.Metadata#updateVersion()}), then
+     * we should check that all of the assignments have a valid position.
+     */
+    private void maybeValidateAssignments() {
```
The name here threw me off because of its generality. I think we need to work metadata into the name to make it clearer. Maybe `validatePositionsOnMetadataChange` or `validatePositionsForLeaderChanges`.

mumrah committed:
PR feedback

mumrah committed:
Remove some streaming code

lbradstreet commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -422,10 +421,13 @@ synchronized int numAssignedPartitions() {
-                .filter(tpState -> isAvailable.test(tpState.topicPartition()) && tpState.value().isFetchable())
-                .map(PartitionStates.PartitionState::topicPartition)
-                .collect(Collectors.toList());
+        List<TopicPartition> result = new ArrayList<>();
```
We should add a small comment that this is in the hotpath and is written the "ugly" way for a reason. It's also probably worth mentioning that we do the cheap isFetchable check first.

mumrah committed:
Add some commentary about ugly code

mumrah committed:
Add JMH for SubscriptionState

mumrah commented:
Here are some results from the new JMH added (note that the units are milliseconds)

```
KAFKA-8806:

Benchmark                                                   (partitionCount)  (topicCount)  Mode  Cnt   Score   Error  Units
SubscriptionStateBenchmark.testFetchablePartitions                        50          5000  avgt   15  13.758 ± 0.477  ms/op
SubscriptionStateBenchmark.testHasAllFetchPositions                       50          5000  avgt   15   9.131 ± 0.056  ms/op
SubscriptionStateBenchmark.testPartitionsNeedingValidation                50          5000  avgt   15  11.869 ± 0.403  ms/op
JMH benchmarks done


Trunk:

Benchmark                                                   (partitionCount)  (topicCount)  Mode  Cnt   Score   Error  Units
SubscriptionStateBenchmark.testFetchablePartitions                        50          5000  avgt   15  19.249 ± 0.117  ms/op
SubscriptionStateBenchmark.testHasAllFetchPositions                       50          5000  avgt   15  17.025 ± 1.421  ms/op
SubscriptionStateBenchmark.testPartitionsNeedingValidation                50          5000  avgt   15  13.291 ± 0.646  ms/op
JMH benchmarks done
```

So for very high partition counts, there seems to be a decent improvement

hachikuji reviewed (approved):
LGTM. I did not review the benchmark at depth, but it looks reasonable. For the future, I think we can consider smarter bookkeeping to avoid the need for these loops in the first place. We can create an index in `SubscriptionState` which is keyed by the state of the partition so so that we do not need a pass to discover the resetting/validating partitions (for example).

For the particular case of validation, I am also looking forward to the improvements that are possible with the Fetch changes in KIP-595. Basically validation can be piggybacked on the Fetch API and we can avoid the need for a separate validating state.

mumrah committed:
Add spotbugs exclude for new jmh package

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -639,26 +643,34 @@ public synchronized OffsetResetStrategy resetStrategy(TopicPartition partition)
+                return false;
+            }
+        }
+        return true;
```
@mumrah I now understand why the previous version was slower, it was allocating a PartitionState instance per element. But we only use the value here. So, we could still use `allMatch` without the performance penalty.
